File: 22-generate-parentheses/generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        stack = []
        res = []
        def backtracking(openP, closeP):
            if openP == closeP == n:
                res.append("".join(stack))
                return

            if openP < n:
                stack.append("(")
                backtracking(openP + 1, closeP)
                stack.pop()
            if closeP < openP:
                stack.append(")")
                backtracking(openP, closeP + 1)
                stack.pop()

        backtracking(0,0)
        return res

File: 22-generate-parentheses/generate-parentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        List<string> result = new List<string>();
        Stack<char> stack = new Stack<char>();

        void backTracking(int openN, int closeN){
            if (openN == n && closeN == n){
                result.Add(new String(stack.Reverse().ToArray()));
                return;
            }
            if (openN < n){
                stack.Push('(');
                backTracking(openN + 1, closeN);
                stack.Pop();
            }
            if (closeN < openN){
                stack.Push(')');
                backTracking(openN, closeN + 1);
                stack.Pop();
            }
        }
        backTracking(0,0);
        return result;
        
    }
}